File: lstm.py
import numpy as np
import tqdm
import json

dataset_path = "datasets/shakespeare/tiny_shakespeare_small.txt"
with open(dataset_path, 'r') as file:
    data = file.read()

#? NOTE: Mirko
# Need to use dict.fromKeys to remove duplicates
# since set() does not maintain the order of its
# elements on each run.
vocab = list(dict.fromkeys(data))
vocab_size = len(vocab)

char_to_idx = {c:i for i, c in enumerate(vocab)}
idx_to_char = {i:c for i, c in enumerate(vocab)}

data_size = len(data)
X_train = data[:-1]
Y_train = data[1:]

def tanh(x: np.ndarray) -> np.ndarray:
    return np.tanh(x)
def dtanh(x: np.ndarray) -> np.ndarray:
    return 1 - np.tanh(x)**2

def sigmoid(x: np.ndarray) -> np.ndarray:
    return 1/(1+np.exp(-x))
def dsigmoid(x: np.ndarray) -> np.ndarray:
    return sigmoid(x)*(1-sigmoid(x))

def softmax(x: np.ndarray) -> np.ndarray:
    exp = np.exp(x)
    return exp/np.sum(exp)

def one_hot_encode(c: str) -> np.ndarray:
    one_hot = np.zeros((vocab_size, 1))
    one_hot[char_to_idx[c]] = 1
    
    return one_hot

def init_weights(input_size: int, output_size: int) -> np.ndarray:
    # Normalized Xavier initialization
    return np.random.uniform(-1, 1, (output_size, input_size)) * np.sqrt(6/(input_size + output_size))

class LSTM:
    def __init__(self, input_size: int, hidden_size: int, output_size: int, learning_rate: float = 1e-3):
        # Hyperparameters
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.output_size = output_size
        self.learning_rate = learning_rate

        # Forget gate weights
        self.W_f = init_weights(input_size, self.hidden_size)
        self.b_f = np.zeros((self.hidden_size, 1))

        # Input gate weights
        self.W_i = init_weights(input_size, self.hidden_size)
        self.b_i = np.zeros((self.hidden_size, 1))

        # Candidate gate weights
        self.W_c = init_weights(input_size, self.hidden_size)
        self.b_c = np.zeros((self.hidden_size, 1))

        # Output gate weights
        self.W_o = init_weights(input_size, self.hidden_size)
        self.b_o = np.zeros((self.hidden_size, 1))

        # Final gate weights
        self.W_y = init_weights(self.hidden_size, output_size)
        self.b_y = np.zeros((output_size, 1))
        
        # Network cache
        self.concat_inputs: dict[int, np.ndarray] = {}

        self.hidden_states: dict[int, np.ndarray] = {-1: np.zeros((self.hidden_size, 1))}
        self.cell_states: dict[int, np.ndarray] = {-1: np.zeros((self.hidden_size, 1))}

        self.forget_gates: dict[int, np.ndarray] = {}
        self.input_gates: dict[int, np.ndarray] = {}
        self.candidate_gates: dict[int, np.ndarray] = {}
        self.output_gates: dict[int, np.ndarray] = {}
        self.activation_outputs: dict[int, np.ndarray] = {}
        self.outputs: dict[int, np.ndarray] = {}
        
    def reset_cache(self):
        self.concat_inputs = {}
        
        self.hidden_states = {-1: np.zeros((self.hidden_size, 1))}
        self.cell_states = {-1: np.zeros((self.hidden_size, 1))}
        
        self.forget_gates = {}
        self.input_gates = {}
        self.candidate_gates = {}
        self.output_gates = {}
        self.activation_outputs = {}
        self.outputs = {}
        
    def forward(self, inputs: list[np.ndarray]) -> list[np.ndarray]:
        self.reset_cache()
        
        outputs = []
        for t in range(len(inputs)):
            self.concat_inputs[t] = np.concatenate((self.hidden_states[t-1], inputs[t]))
            
            self.forget_gates[t] = np.dot(self.W_f, self.concat_inputs[t]) + self.b_f
            self.input_gates[t] = np.dot(self.W_i, self.concat_inputs[t]) + self.b_i
            self.candidate_gates[t] = np.dot(self.W_c, self.concat_inputs[t]) + self.b_c
            self.output_gates[t] = np.dot(self.W_o, self.concat_inputs[t]) + self.b_o
            
            fga = sigmoid(self.forget_gates[t])
            iga = sigmoid(self.input_gates[t])
            cga = tanh(self.candidate_gates[t])
            oga = sigmoid(self.output_gates[t])
            
            self.cell_states[t] = fga * self.cell_states[t-1] + iga * cga
            self.hidden_states[t] = oga * tanh(self.cell_states[t])
            
            outputs += [np.dot(self.W_y, self.hidden_states[t]) + self.b_y]
            
        return outputs
    
    def backward(self, errors: list[np.ndarray]) -> None:
        dW_f, db_f = 0, 0
        dW_i, db_i = 0, 0
        dW_c, db_c = 0, 0
        dW_o, db_o = 0, 0
        dW_y, db_y = 0, 0
        
        hidden_state = np.zeros_like(self.hidden_states[0])
        cell_state = np.zeros_like(self.cell_states[0])
        
        for t in reversed(range(len(self.concat_inputs))):
            err = errors[t]
            dW_y += np.dot(err, self.hidden_states[t].T)
            db_y += err
            
            dhidden = np.dot(self.W_y.T, err) + hidden_state
            doutput = tanh(self.cell_states[t]) * dhidden * dsigmoid(self.output_gates[t])
            
            dW_o += np.dot(doutput, self.concat_inputs[t].T)
            db_o += doutput
            
            dcell_state = dtanh(self.cell_states[t]) * sigmoid(self.output_gates[t]) * dhidden + cell_state
            
            dforget = dcell_state * self.cell_states[t-1] * dsigmoid(self.forget_gates[t])
            
            dW_f += np.dot(dforget, self.concat_inputs[t].T)
            db_f += dforget
            
            dinput = dcell_state * tanh(self.candidate_gates[t]) * dsigmoid(self.input_gates[t])
            
            dW_i += np.dot(dinput, self.concat_inputs[t].T)
            db_i += dinput
            
            dcandidate = dcell_state * sigmoid(self.input_gates[t]) * dtanh(self.candidate_gates[t])
            
            dW_c += np.dot(dcandidate, self.concat_inputs[t].T)
            db_c += dcandidate
            
            d_concat_inputs = np.dot(self.W_f.T, dforget) + np.dot(self.W_i.T, dinput) + np.dot(self.W_c.T, dcandidate) + np.dot(self.W_o.T, doutput)
            
            hidden_state = d_concat_inputs[:self.hidden_size, :]
            cell_state = sigmoid(self.forget_gates[t]) * dcell_state
            
        for _d in (dW_f, db_f, dW_i, db_i, dW_c, db_c, dW_o, db_o, dW_y, db_y):
            np.clip(_d, -1, 1, out=_d)
            
        self.W_f -= dW_f * self.learning_rate
        self.b_f -= db_f * self.learning_rate
        
        self.W_i -= dW_i * self.learning_rate
        self.b_i -= db_i * self.learning_rate
        
        self.W_c -= dW_c * self.learning_rate
        self.b_c -= db_c * self.learning_rate
        
        self.W_o -= dW_o * self.learning_rate
        self.b_o -= db_o * self.learning_rate
        
        self.W_y -= dW_y * self.learning_rate
        self.b_y -= db_y * self.learning_rate
        
    def train(self, inputs: list[str], labels: list[str], epochs: int = 1):
        one_hot_inputs = [one_hot_encode(c) for c in inputs]
        
        for _ in tqdm.tqdm(range(epochs)):
            predictions = self.forward(one_hot_inputs)
                
            errors = []
            for i in range(len(predictions)):
                errors += [softmax(predictions[i])]
                errors[-1][char_to_idx[labels[i]]] -= 1

            self.backward(errors)
    
    def save_weights(self, weights_path: str) -> None:
        weights = {
            'W_f': self.W_f.tolist(),
            'b_f': self.b_f.tolist(),
            'W_i': self.W_i.tolist(),
            'b_i': self.b_i.tolist(),
            'W_c': self.W_c.tolist(),
            'b_c': self.b_c.tolist(),
            'W_o': self.W_o.tolist(),
            'b_o': self.b_o.tolist(),
            'W_y': self.W_y.tolist(),
            'b_y': self.b_y.tolist(),
        }
            
        
        with open(weights_path, 'w') as f:
            json.dump(weights, f)
    
    def load_weights(self, weights_path: str) -> None:
        with open(weights_path, 'r') as f:
            weights = json.load(f)
            
            self.W_f = np.array(weights['W_f'])
            self.b_f = np.array(weights['b_f'])
            
            self.W_i = np.array(weights['W_i'])
            self.b_i = np.array(weights['b_i'])
            
            self.W_c = np.array(weights['W_c'])
            self.b_c = np.array(weights['b_c'])
            self.W_o = np.array(weights['W_o'])
            self.b_o = np.array(weights['b_o'])
            
            self.W_y = np.array(weights['W_y'])
            self.b_y = np.array(weights['b_y'])
    
    def test(self, inputs: list[str], labels: list[str]):
        accuracy = 0
        probabilities = self.forward([one_hot_encode(input) for input in inputs])

        output = ''
        for i in range(len(labels)):
            prediction = idx_to_char[np.random.choice(vocab_size, p = softmax(probabilities[i].reshape(-1)))]

            output += prediction

            if prediction == labels[i]:
                accuracy += 1

        # print(f'Predictions:\nt{"".join(output)}\n')
        print(f'Accuracy: {round(accuracy * 100 / len(inputs), 2)}%')

import argparse

if __name__ == "__main__":
    hidden_size = 64
    input_size = vocab_size + hidden_size
    output_size = vocab_size

    learning_rate = 0.05
    epochs = 10
    
    lstm = LSTM(input_size, hidden_size, output_size, learning_rate)
    
    parser = argparse.ArgumentParser()
    parser.add_argument("--mode", type=str, help="LSTM mode: train or test")
    parser.add_argument("--weights", type=str, help="Weights file path")

    args = parser.parse_args()
    weights_path = args.weights if args.weights else "./weights/lstm_weights.json"
    mode = args.mode if args.mode else "train"

    print(f"Running LSTM network in {mode} mode...")

    if mode == 'train':
        lstm.train(X_train, Y_train, epochs)
        lstm.save_weights(weights_path)
    elif mode == 'test':
        lstm.load_weights(weights_path)
        lstm.test(X_train, Y_train)
    elif mode == 'optimize':
        lstm.load_weights(weights_path)
        lstm.train(X_train, Y_train, epochs=10)
        lstm.save_weights(weights_path)
    else:
        print("Invalid mode. Use 'train', 'test' or 'optimize'.")

File: cpp/lstm.cpp
#include "lstm.hpp"

#include <cassert>
#include <cmath>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <nlohmann/json.hpp>

static Matrix dtanh(const Matrix &x) {
    return 1.0f - x.tanh().pow(2.0f);
}

static Matrix dsigmoid(const Matrix &x) {
    return x.sigmoid() * (1.0f - x.sigmoid());
}

static Matrix init_weights(uint32_t input_size, uint32_t output_size) {
    // Xavier initialization
    return Matrix::uniform(output_size, input_size, -1.0f, 1.0f) *
           std::sqrt(6.0f / (input_size + output_size));
}

LSTM::LSTM(const LSTM &other)
    : input_size(other.input_size), hidden_size(other.hidden_size), output_size(other.output_size),
      learning_rate(other.learning_rate), W_f(other.W_f), b_f(other.b_f), W_i(other.W_i),
      b_i(other.b_i), W_c(other.W_c), b_c(other.b_c), W_o(other.W_o), b_o(other.b_o),
      W_y(other.W_y), b_y(other.b_y) {}

LSTM::LSTM(uint32_t input_size, uint32_t hidden_size, uint32_t output_size, float learning_rate)
    : input_size(input_size), hidden_size(hidden_size), output_size(output_size),
      learning_rate(learning_rate) {
    W_f = init_weights(input_size, hidden_size);
    b_f = Matrix::zeros(hidden_size, 1);

    W_i = init_weights(input_size, hidden_size);
    b_i = Matrix::zeros(hidden_size, 1);

    W_c = init_weights(input_size, hidden_size);
    b_c = Matrix::zeros(hidden_size, 1);

    W_o = init_weights(input_size, hidden_size);
    b_o = Matrix::zeros(hidden_size, 1);

    W_y = init_weights(hidden_size, output_size);
    b_y = Matrix::zeros(output_size, 1);

    reset_cache();
}

void LSTM::reset_cache() {
    concat_inputs.clear();

    hidden_states.clear();
    hidden_states[-1] = Matrix::zeros(hidden_size, 1);
    cell_states.clear();
    cell_states[-1] = Matrix::zeros(hidden_size, 1);

    forget_gates.clear();
    input_gates.clear();
    candidate_gates.clear();
    output_gates.clear();
    activation_outputs.clear();
    outputs.clear();
}

std::vector<Matrix> LSTM::forward(const std::vector<Matrix> &inputs) {
    reset_cache();

    std::vector<Matrix> ret_outputs;
    ret_outputs.reserve(inputs.size());

    for (int32_t t = 0; t < static_cast<int32_t>(inputs.size()); t++) {
        concat_inputs[t] = Matrix::concatenate(0, {hidden_states[t - 1], inputs[t]});
        forget_gates[t] = W_f.matmul(concat_inputs[t]) + b_f;
        input_gates[t] = W_i.matmul(concat_inputs[t]) + b_i;
        candidate_gates[t] = W_c.matmul(concat_inputs[t]) + b_c;
        output_gates[t] = W_o.matmul(concat_inputs[t]) + b_o;

        Matrix fga = forget_gates[t].sigmoid();
        Matrix iga = input_gates[t].sigmoid();
        Matrix cga = candidate_gates[t].tanh();
        Matrix oga = output_gates[t].sigmoid();

        cell_states[t] = fga * cell_states[t - 1] + iga * cga;
        hidden_states[t] = oga * cell_states[t].tanh();

        Matrix output = (W_y.matmul(hidden_states[t]) + b_y).softmax(0);
        ret_outputs.push_back(output);
        outputs[t] = output;
    }

    return ret_outputs;
}

void LSTM::backward(const std::vector<Matrix> &labels) {
    Matrix dW_f = Matrix::zeros_like(W_f);
    Matrix db_f = Matrix::zeros_like(b_f);
    Matrix dW_i = Matrix::zeros_like(W_i);
    Matrix db_i = Matrix::zeros_like(b_i);
    Matrix dW_c = Matrix::zeros_like(W_c);
    Matrix db_c = Matrix::zeros_like(b_c);
    Matrix dW_o = Matrix::zeros_like(W_o);
    Matrix db_o = Matrix::zeros_like(b_o);
    Matrix dW_y = Matrix::zeros_like(W_y);
    Matrix db_y = Matrix::zeros_like(b_y);

    Matrix hidden_state = Matrix::zeros_like(hidden_states[0]);
    Matrix cell_state = Matrix::zeros_like(cell_states[0]);

    for (int32_t t = static_cast<int32_t>(concat_inputs.size() - 1); t >= 0; t--) {
        Matrix dL = -(labels[t] / outputs[t]);
        Matrix dsoftmax =
            outputs[t] * dL -
            outputs[t].T().matmul(dL).expand(0, outputs[t].shape().first) * outputs[t];

        dW_y = dW_y + dsoftmax.matmul(hidden_states[t].T());
        db_y = db_y + dsoftmax;

        Matrix dhidden_state = W_y.T().matmul(dsoftmax) + hidden_state;
        Matrix doutput = cell_states[t].tanh() * dhidden_state * dsigmoid(output_gates[t]);

        dW_o = dW_o + doutput.matmul(concat_inputs[t].T());
        db_o = db_o + doutput;

        Matrix dcell_state =
            dtanh(cell_states[t]) * output_gates[t].sigmoid() * dhidden_state + cell_state;

        Matrix dforget = dcell_state * cell_states[t - 1] * dsigmoid(forget_gates[t]);

        dW_f = dW_f + dforget.matmul(concat_inputs[t].T());
        db_f = db_f + dforget;

        Matrix dinput = dcell_state * candidate_gates[t].tanh() * dsigmoid(input_gates[t]);

        dW_i = dW_i + dinput.matmul(concat_inputs[t].T());
        db_i = db_i + dinput;

        Matrix dcandidate = dcell_state * input_gates[t].sigmoid() * dtanh(candidate_gates[t]);

        dW_c = dW_c + dcandidate.matmul(concat_inputs[t].T());
        db_c = db_c + dcandidate;

        Matrix dconcat_inputs = W_f.T().matmul(dforget) + W_i.T().matmul(dinput) +
                                W_c.T().matmul(dcandidate) + W_o.T().matmul(doutput);

        uint32_t shrink_size = dconcat_inputs.shape().first - hidden_size;
        hidden_state = dconcat_inputs.shrink_end(0, shrink_size);
        cell_state = forget_gates[t].sigmoid() * dcell_state;
    }

    // Clip gradients to prevent exploding
    dW_f = dW_f.clamp(-1.0f, 1.0f);
    db_f = db_f.clamp(-1.0f, 1.0f);
    dW_i = dW_i.clamp(-1.0f, 1.0f);
    db_i = db_i.clamp(-1.0f, 1.0f);
    dW_c = dW_c.clamp(-1.0f, 1.0f);
    db_c = db_c.clamp(-1.0f, 1.0f);
    dW_o = dW_o.clamp(-1.0f, 1.0f);
    db_o = db_o.clamp(-1.0f, 1.0f);
    dW_y = dW_y.clamp(-1.0f, 1.0f);
    db_y = db_y.clamp(-1.0f, 1.0f);

    W_f = W_f - dW_f * learning_rate;
    b_f = b_f - db_f * learning_rate;
    W_i = W_i - dW_i * learning_rate;
    b_i = b_i - db_i * learning_rate;
    W_c = W_c - dW_c * learning_rate;
    b_c = b_c - db_c * learning_rate;
    W_o = W_o - dW_o * learning_rate;
    b_o = b_o - db_o * learning_rate;
    W_y = W_y - dW_y * learning_rate;
    b_y = b_y - db_y * learning_rate;
}

void LSTM::save(const std::string &model_path) const {
    std::filesystem::path fp(model_path);
    std::filesystem::path dir_path = fp.parent_path();

    if (!std::filesystem::exists(dir_path)) {
        std::filesystem::create_directories(dir_path);
    }

    std::ofstream ofs(model_path);

    if (!ofs.is_open()) {
        throw std::runtime_error("Failed to open file for writing");
    }

    nlohmann::json json_obj;

    json_obj["input_size"] = input_size;
    json_obj["hidden_size"] = hidden_size;
    json_obj["output_size"] = output_size;

    json_obj["learning_rate"] = learning_rate;

    json_obj["W_f"] = W_f.get_data();
    json_obj["b_f"] = b_f.get_data();

    json_obj["W_i"] = W_i.get_data();
    json_obj["b_i"] = b_i.get_data();

    json_obj["W_c"] = W_c.get_data();
    json_obj["b_c"] = b_c.get_data();

    json_obj["W_o"] = W_o.get_data();
    json_obj["b_o"] = b_o.get_data();

    json_obj["W_y"] = W_y.get_data();
    json_obj["b_y"] = b_y.get_data();

    ofs << json_obj.dump(4) << std::endl;
    ofs.close();
}

LSTM LSTM::load(const std::string &model_path) {
    std::ifstream ifs(model_path);

    if (!ifs.is_open()) {
        throw std::runtime_error("Failed to open file for reading");
    }

    nlohmann::json json_obj;
    ifs >> json_obj;
    ifs.close();

    uint32_t input_size = json_obj["input_size"];
    uint32_t hidden_size = json_obj["hidden_size"];
    uint32_t output_size = json_obj["output_size"];

    float learning_rate = json_obj["learning_rate"];

    LSTM lstm(input_size, hidden_size, output_size, learning_rate);

    // 2D arrays will be converted to std::vector<std::vector<float>>
    lstm.W_f = Matrix(json_obj["W_f"]);
    lstm.b_f = Matrix(json_obj["b_f"]);

    lstm.W_i = Matrix(json_obj["W_i"]);
    lstm.b_i = Matrix(json_obj["b_i"]);

    lstm.W_c = Matrix(json_obj["W_c"]);
    lstm.b_c = Matrix(json_obj["b_c"]);

    lstm.W_o = Matrix(json_obj["W_o"]);
    lstm.b_o = Matrix(json_obj["b_o"]);

    lstm.W_y = Matrix(json_obj["W_y"]);
    lstm.b_y = Matrix(json_obj["b_y"]);

    return std::move(lstm);
}
